cpp/src/bitmask/null_mask.cu

```diff
@@ -128,8 +145,92 @@ CUDF_KERNEL void set_null_mask_kernel(bitmask_type* __restrict__ destination,
     }
   }
 }
+
+CUDF_KERNEL void set_null_mask_bulk_kernel(
+  cudf::device_span<bitmask_type* __restrict__> destinations,
+  cudf::device_span<size_type const> begin_bits,
+  cudf::device_span<size_type const> end_bits,
+  cudf::device_span<bool const> valids,
+  cudf::device_span<size_type const> numbers_of_mask_words)
+{
+  auto const bitmask_idx = cg::this_grid().block_rank();
+  // Return early if nothing to do
+  if (begin_bits[bitmask_idx] == end_bits[bitmask_idx]) { return; }
+  set_null_mask_impl(destinations[bitmask_idx],
+                     begin_bits[bitmask_idx],
+                     end_bits[bitmask_idx],
+                     valids[bitmask_idx],
+                     numbers_of_mask_words[bitmask_idx],
+                     cg::this_thread_block());
+}
+
+CUDF_KERNEL void set_null_mask_kernel(bitmask_type* __restrict__ destination,
+                                      size_type begin_bit,
+                                      size_type end_bit,
+                                      bool valid,
+                                      size_type number_of_mask_words)
+{
+  set_null_mask_impl(destination, begin_bit, end_bit, valid, number_of_mask_words, cg::this_grid());
+}
 }  // namespace
 
+// Set pre-allocated null masks of given bit ranges [begin_bit, end_bit) to valids, if valid==true,
+// or null, otherwise;
+void set_null_masks_bulk(cudf::host_span<bitmask_type*> bitmasks,
+                         cudf::host_span<size_type const> begin_bits,
+                         cudf::host_span<size_type const> end_bits,
+                         cudf::host_span<bool const> valids,
+                         rmm::cuda_stream_view stream)
+{
+  CUDF_FUNC_RANGE();
+
+  auto const num_bitmasks = bitmasks.size();
+
+  CUDF_EXPECTS(num_bitmasks == begin_bits.size(),
+               "Number of bitmasks and begin bits must be equal.");
+  CUDF_EXPECTS(num_bitmasks == end_bits.size(), "Number of bitmasks and end bits must be equal.");
+  CUDF_EXPECTS(num_bitmasks == valids.size(), "Number of bitmasks and valids must be equal.");
+
+  size_t average_nullmask_words = 0;
+  auto h_number_of_mask_words   = cudf::detail::make_host_vector<size_type>(num_bitmasks, stream);
+  thrust::tabulate(
+    thrust::host, h_number_of_mask_words.begin(), h_number_of_mask_words.end(), [&](auto i) {
+      CUDF_EXPECTS(begin_bits[i] >= 0, "Invalid range.");
+      CUDF_EXPECTS(begin_bits[i] <= end_bits[i], "Invalid bit range.");
+      // Return 0 if bitmask is empty
+      if (begin_bits[i] == end_bits[i]) { return size_t{0}; }
+      // Number of words in this bitmask
+      auto const num_words =
+        num_bitmask_words(end_bits[i]) - begin_bits[i] / detail::size_in_bits<bitmask_type>();
+      // Divide by num_bitmasks and update average here to avoid overflow
+      average_nullmask_words += cudf::util::div_rounding_up_safe<size_t>(num_words, num_bitmasks);
```

Comment:
I guess overflow is a real enough issue to be worth injecting this many divisions? If it mattered we could use an implementation that keeps track of the overflows, but that's probably a premature optimization I assume?

Reply:
Saw the overflow in one of the benchmarks. I will add a more clever way to reduce the number of divisions